_posts/pl/2021-02-10-przyklad-migracji-do-chmury.md

```diff
@@ -0,0 +1,888 @@
+
+Migracja strategią "Rip and replace" polega na całkowitym wykorzystaniu natywnych rozwiązań chmury. Oprócz usług SaaS, wykorzystujemy dodatkowo Cloud Functions, dzięki czemu eliminujemy minusy poprzednich dwóch strategii.
+
+![Architektura aplikacji docelowej - "Rip and replace"](/assets/img/posts/2021-02-10-przyklad-migracji-do-chmury/rip_and_replace.png)
+
+**Plusy:**
+Optymalizujemy architekturę systemu, przesuwając granicę odpowiedzialności prawie całkowicie na stronę usługodawcy. Cloud Functions umożliwia wywoływanie funkcji po pojawieniu się wiadomości w usłudze Cloud Pub/Sub, dzięki czemu aplikacja frontendowa może zapisywać dane przyjęte od użytkownika bezpośrednio do kolejki, bez pośrednictwa aplikacji backendowej. Przepisując aplikację backendową na funkcję można zrezygnować również z jej pośrednictwa w odczycie danych z Cloud Firestore, ponieważ usługa ta udostępnia REST API, które umożliwia odczyt danych bezpośrednio przez aplikację frontendową. Dzięki automatycznemu skalowaniu (również do zera), optymalizujemy koszty utrzymania infrastruktury, które w przypadku zerowego ruchu użytkowników, można ograniczyć całkowicie do zera (w ramach [GCP Free Tier](https://cloud.google.com/free)).
+
+**Minusy:**
+Wykorzystanie natywnych rozwiązań wymaga od nas bardzo dobrej znajomości platformy, usług oraz komunikacji między nimi. Jak okaże się za chwilę, pojawiają się również nowe problemy, jak na przykład ograniczenia API usług, czy problem zimnego startu funkcji. Migracja z wykorzystaniem strategii "Rip and replace" jest również najbardziej czasochłonna, ponieważ wymaga całkowitego przepisania części systemu.
+
+## Architektura systemu docelowego
+
+Ponieważ złożoność systemu źródłowego jest niewielka i chciałbym w tym wpisie pokazać jak najwięcej, przeprowadzimy migrację z wykorzystaniem strategii "Rip and replace". Architektura systemu docelowego będzie więc wyglądać następująco:
+
+- **aplikacja frontendowa** (pliki statyczne) zostanie umieszczona w Cloud Storage i udostępniona publicznie, wykorzystując [HTTP/S Load Balancer](https://cloud.google.com/load-balancing/docs/https). Alternatywą jest wykorzystanie platformy [Firebase](https://firebase.google.com/), w praktyce jednak, Firebase przechowuje pliki statyczne również w Cloud Storage, przykrywając wszystko własnym interfejsem graficznym;
+- **aplikacja backendowa**, z racji niskiej złożoności, zostanie przepisana całkowicie na Cloud Functions. Utracona zostanie walidacja danych wejściowych, ale dzięki temu użytkownik będzie mógł przesyłać dane do Cloud Pub/Sub bezpośrednio z aplikacji frontendowej. Dane te mogą być walidowane przez funkcję, która pobiera je z Cloud Pub/Sub, przetwarza i utrwala w Firestore. Gdyby walidacja była wymagana przed wrzuceniem danych do Cloud Pub/Sub, to pomiędzy aplikacją frontendową a Cloud Pub/Sub musiałaby znaleźć się dodatkowa funkcja, której zadaniem byłoby przyjęcie lub odrzucenie danych oraz przesłanie ich do Cloud Pub/Sub;
```

Comment:
a nie chcelibyśmy jednak, żeby system po migracji działał tak samo? Bo to jest istotne uproszczenie...

Reply:
Chciałem pokazać, że jednym z etapów migracji jest również refactoring i upraszczanie architektury. Żeby nie było, że system jest uproszczony tylko dla ułatwienia migracji, opisałem w jaki sposób można zrealizować system tak, aby działał tak samo jak przed migracją.